pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// Edit this file to define custom logic or remove it if it is not needed.
+/// Learn more about FRAME and the core library of Substrate FRAME pallets:
+/// <https://substrate.dev/docs/en/knowledgebase/runtime/frame>
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+use codec::HasCompact;
+use common_traits::{PoolNAV, PoolReserve};
+use core::{convert::TryFrom, ops::AddAssign};
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*, traits::UnixTime};
+use frame_system::pallet_prelude::*;
+use orml_traits::MultiCurrency;
+use sp_runtime::{
+	traits::{
+		AccountIdConversion, AtLeast32BitUnsigned, CheckedAdd, CheckedMul, CheckedSub, One,
+		Saturating, Zero,
+	},
+	FixedPointNumber, FixedPointOperand, PerThing, Perquintill, TypeId,
+};
+use sp_std::vec::Vec;
+
+/// Trait for converting a pool+tranche ID pair to a CurrencyId
+///
+/// This should be implemented in the runtime to convert from the
+/// PoolId and TrancheId types to a CurrencyId that represents that
+/// tranche.
+///
+/// The pool epoch logic assumes that every tranche has a UNIQUE
+/// currency, but nothing enforces that. Failure to ensure currency
+/// uniqueness will almost certainly cause some wild bugs.
+pub trait TrancheToken<T: Config> {
+	fn tranche_token(pool: T::PoolId, tranche: T::TrancheId) -> T::CurrencyId;
+}
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, Default)]
+pub struct Tranche<Balance> {
+	pub interest_per_sec: Perquintill,
```

Comment:
Perquintill _is_ a fixed point type, technically. It's the semantically-correct type for a number which cannot exceed 1.0.

This only stores the base interest rate and is not used as an accumulator. Intermediate values for interest calculations do use 27 decimals. I think 18 decimals is sufficient here, but if we really want to store 27 we should make a `Peroctill` type

Reply:
I think we should use the 27 precision

Reply:
That's fine. I'll add an item to the things-to-do in December to add a `Peroctill` we can use for 27-decimal proper fractions and update this. Should be a one-liner in our substrate fork and here.

Reply:
We're already still using a fork for the other 27-digit fixed points, right? Or did we finally get that upstream?

I may actually use Fixed27 instead of Peroctill, so that the interest is in the form of `1.<rate>`. it saves one addition for every interest calculation

Reply:
We ware using the main repo on parachain at least. Would prefer to keep it that way 😄

Reply:
We dont use the fork anymore. I have moved the macro to our repo. You can already one with it